src/macaron/slsa_analyzer/git_url.py

```diff
@@ -25,22 +25,126 @@
+    remote : str, optional
+        The name of the remote to check the branches, by default "origin".
+
+    Returns
+    -------
+    list[str]
+        The list of branches that contains the commit.
+    """
+    try:
+        raw_output: str = git_obj.repo.git.branch(
+            "--remotes",
+            "--list",
+            f"{remote}/*",
+            "--contains",
+            commit,
+        )
+    except GitCommandError:
+        logger.debug("Error while using git branch to look for branches that contains %s.", commit)
+        return []
+
+    return parse_git_branch_output(raw_output)
+
+
+def check_out_repo_target(
+    git_obj: Git,
+    branch_name: str = "",
+    digest: str = "",
+    offline_mode: bool = False,
+) -> bool:
     """Checkout the branch and commit specified by the user.
 
-    If no branch name is provided, this method will checkout the default branch
-    of the repository and analyze the latest commit from remote. Note that checking out the branch
-    is always performed before checking out the specific ``digest`` (if provided).
+    This fucntion assumes that a remote "origin" exist and checkout from that remote ONLY.
 
-    If ``digest`` is not provided, this method always pulls (fast-forward only) and checks out the latest commit.
+    If ``offline_mode`` is False, this function will fetch new changes from origin remote. The fetching operation
+    will prune and update all references (e.g. tags, branches) to make sure that the local repository is up-to-date
+    with the repository specified by origin remote.
 
-    If ``digest`` is provided, this method will checkout that specific commit. If ``digest``
-    cannot be found in the current branch, this method will pull (fast-forward only) from remote.
+    If ``branch_name`` and a commit are not provided, this function will checkout the latest commit of the
+    default branch (i.e. origin/HEAD).
 
-    This method supports repositories which are cloned from existing remote repositories.
-    Other scenarios are not covered (e.g. a newly initiated repository).
+    If ``branch_name`` is provided and a commit is not provided, this function will checkout that branch from origin
+    remote (i.e. origin/<branch_name).
+
+    If ``branch_name`` is not provided and a commit is provided, this function will checkout the commit directly.
 
-    If ``offline_mode`` is set, this method will not pull/fetch from remote while checking out the branch or commit.
+    If both ``branch_name`` and a commit are provided, this function will checkout the commit directly only if that
+    commit exists in the branch origin/<branch_name>. If not, this fucntion will return False.
+
+    For all scenarios:
+    - If the checkout failed (e.g. a branch or a commit doesn't exist), this function will return
+    False.
+    - This function will perform a force checkout, which mean it will reset the index and working tree of the target
```

Comment:
Here I think:
* by "index" you probably mean "staging area"
* by "working tree" you probably mean "working directory"

I assume you are referring to the 2 in 3 states of files in git, including "working directory" (where unstaged changes live), "staging area"/"index" (where staged changes live), and "commits". For reference, this is similar to what [this Git documentation page describes](https://git-scm.com/book/en/v2/Getting-Started-What-is-Git%3F).

I think "index" is fine, just a bit less unintuitive than "staging area" so I would personally prefer "staging area", but I don't have a strong opinion about it.
However, "working tree" is probably not a term to use here because it can have a different meaning in the context of Git. It may refer to [Git worktree](https://git-scm.com/docs/git-worktree/2.24.0) instead, which is a different concept and probably not what you are referring to here.

Reply:
I agree that using `staging area` is better than `index`. I will update it.

With `working tree`, it's still correct with my intention of referring to the files checked out from the git repository at a commit to the file system:

> The working tree is a single checkout of one version of the project. These files are pulled out of the compressed database in the Git directory and placed on disk for you to use or modify.

In the documentation page you mentioned, `working tree` is used more frequently than `working directory` (even though they should shared the same definition). That's why I thought `working tree` would be a better choice. This is also consistent with the usage of `working tree` in other git command documentation too. For example: 
- https://git-scm.com/docs/git-reset
- This [post](https://stackoverflow.com/questions/39128500/working-tree-vs-working-directory) about `working tree` vs `working directory` in `git status`
  - i.e  They claimed that `working directory` can mean the current directory where a git command is run instead of the root directory of the checked out git repo files.

In the documentation of [Git worktree](https://git-scm.com/docs/git-worktree/2.24.0), I think `working tree` still keeps the same meaning as above because Git worktree allows us to manage different working trees (each working tree is checkout from a branch for example). 

I think what we can do is instead of mentioning the `index/staging are` and `working tree/working directory` in the doc string, we could reference to the documentation of git checkout --force ([here](https://git-scm.com/docs/git-checkout#Documentation/git-checkout.txt---force)) which will ensure one source of truth and prevent any further confusion? :thinking:

Reply:
> In the documentation of [Git worktree](https://git-scm.com/docs/git-worktree/2.24.0), I think working tree still keeps the same meaning as above

After re-reading, I think I agree. I think it was just me mixing up the meaning of "working tree" with the `git worktree` feature 😓. Thanks for verifying, Nhan.

I can see that the meaning of "working tree" should be the same as "working directory", because "trees" in Git refer to "directories". Whether the word "tree" is confusing or not I guess depends on how familiar the reader to the tree concept in Git.

With that, I don't have a strong opinion on whether a "working tree" or "working directory" should be used. I think as long as we are consistent with how we use these terms, then it's fine for me.

Reply:
Thanks Nathan. I would prefer going with working tree to be consistent with how it's frequently used in the documentation. As for index/staging area, I would go with staging area. 
However, in this pydoc string, I think it's best if I reference the documentation of the `--force` flag for checkout ([here](https://git-scm.com/docs/git-checkout#Documentation/git-checkout.txt---force)) instead of mentioning the terminologies.

Reply:
Addressed in https://github.com/oracle/macaron/pull/548/commits/7ee578fd355d5fbaba745e6ac5ada110896bc957